src/osp/tasks/top_utils.h

```diff
@@ -0,0 +1,183 @@
+{
+    while ( bool(topData[current]) && (current < topData.size()) )
+    {
+        ++current;
+    }
+
+    if (current < topData.size())
+    {
+        topData[current].emplace<Reserved>();
+    }
+
+    return current;
+}
+
+template <typename IT_T, typename ITB_T>
+TopDataId top_reserve(
+        ArrayView<entt::any> const topData, TopDataId current,
+        IT_T destFirst, const ITB_T destLast)
+{
+    while (destFirst != destLast)
+    {
+        current = top_reserve(topData, current);
+
+        if (current < topData.size())
+        {
+            *destFirst = current;
+            ++current;
+            std::advance(destFirst, 1);
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    return current;
+}
+
+template <typename T, typename ... ARGS_T>
+T& top_emplace(ArrayView<entt::any> const topData, TopDataId id, ARGS_T&& ... args)
+{
+    entt::any &rData = topData[std::size_t(id)];
+    rData.emplace<T>(std::forward<ARGS_T>(args) ...);
+    return entt::any_cast<T&>(rData);
+}
+
+template <typename T, typename ANY_T>
+T& top_assign(ArrayView<entt::any> const topData, TopDataId id, ANY_T&& any)
+{
+    entt::any &rData = topData[std::size_t(id)];
+    rData = std::forward<ANY_T>(any);
+    return entt::any_cast<T&>(rData);
+}
+
+
+template <typename T>
+[[nodiscard]] T& top_get(ArrayView<entt::any> const topData, TopDataId const id)
+{
+    return entt::any_cast<T&>(topData[id]);
+}
+
+
+
+template<typename FUNCTOR_T>
+struct wrap_args_trait
+{
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr void apply_void(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper_void([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        apply_void<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+
+        return TopTaskStatus::Success;
+    }
+
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr TopTaskStatus apply(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        return FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        return apply<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+    }
+
+    template<typename ... ARGS_T>
+    static constexpr TopTaskFunc_t unpack([[maybe_unused]] TopTaskStatus(*func)(ARGS_T...))
```

Comment:
it appears that all of these only ever work with noexcept functions.

if so, mark the functions that you're taking as parameters as noexcept so that no one gives you an exception throwing function.